### ISSUE TITLE
Schema generation for subclasses with generics

### ISSUE THREAD

**Issue by bjonnhy** [micronaut-projects/micronaut-openapi#259]
### Task List

- [x] Steps to reproduce provided
- [ ] Stacktrace (if present) provided
- [x] Example that reproduces the problem uploaded to Github
- [x] Full description of the issue provided (see below)

### Steps to Reproduce

1. Clone referenced repo
2. Run ./gradlew openapiCopy
3. Open __projectRoot__/build/openapi/swagger/api*.yml

### Expected Behaviour

I'd expect the generated specs to look like:
```yml
openapi: 3.0.1
info:
  title: api
  description: API
  version: "0.01"
servers:
- url: http://dev.lan
paths:
  /time:
    get:
      summary: Time of 10 seconds
      description: Time of 10 seconds
      operationId: getTime
      parameters: []
      responses:
        default:
          description: Time of 10 seconds
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/Time'
components:
  schemas:
    TimeUnit:
      type: string
      enum:
        - Millisecond
        - Second
        - Minute
        - Hour
        - Day
        - Week
    Time:
      type: object
      properties:
        amount:
          type: number
        unit:
          $ref: '#/components/schemas/TimeUnit'   
```

### Actual Behaviour

```yml
openapi: 3.0.1
info:
  title: api
  description: API
  version: "0.01"
servers:
- url: http://dev.lan
paths:
  /time:
    get:
      summary: Time of 10 seconds
      description: Time of 10 seconds
      operationId: getTime
      parameters: []
      responses:
        default:
          description: Time of 10 seconds
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/Time'
components:
  schemas:
    Quantity<Quantity,Unit>:
      type: object
      properties:
        amount:
          type: number
        unit:
          $ref: '#/components/schemas/Unit'
    Unit:
      type: object
    Time:
      type: object
      allOf:
      - $ref: '#/components/schemas/Quantity<Quantity,Unit>'
  securitySchemes:
    basicAuth:
      type: http
      scheme: basic
```
When the suberclass defines one of its members as generic binded by an interface, the scema of any subclass seems to ignore the designated generic.

Additionally, the generated openApi spec produces errors on the `Quantity<>` component name: `Component names can only contain the characters A-Z a-z 0-9 - . _`

### Environment Information

- **Operating System**: Ubuntu 20.04
- **Micronaut Version:** 1.3.6
- **Micronaut OpenApi Version:** 1.5.1
- **JDK Version:** 8.0.252

### Example Application

- https://github.com/bjonnhy/micronaut-openapi-schema-example

**Comment by croudet** [comment]
> Additionally, the generated openApi spec produces errors on the Quantity<> component name: > Component names can only contain the characters A-Z a-z 0-9 - . _

Can you tell me what (swagger-ui?) triggers the error?

**Comment by croudet** [comment]
Your `TimeUnit` class is not a bean. It must conform to the bean spec (getter and setter, fluent or not.)

**Comment by bjonnhy** [comment]
> > Additionally, the generated openApi spec produces errors on the Quantity<> component name: > Component names can only contain the characters A-Z a-z 0-9 - . _
> 
> Can you tell me what (swagger-ui?) triggers the error?

If you paste the definition in https://editor.swagger.io/, it shows the error

> Your `Quantity` class is not a bean. It must conform to the bean spec (getter and setter, fluent or not.)

Changed to `get*` getters right after you posted this comment. Sorry about that.

I also (just now) implemented Serializable, some dummy setters on `Quantity` and override those on `Time`. It seems the overridden getter on `Time` gives some improvements to `amount`, but `unit` still seems incomplete:
```
openapi: 3.0.1
info:
  title: api
  description: API
  version: "0.01"
servers:
- url: http://dev.lan
paths:
  /time:
    get:
      summary: Time of 10 seconds
      description: Time of 10 seconds
      operationId: getTime
      parameters: []
      responses:
        default:
          description: Time of 10 seconds
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/Time'
components:
  schemas:
    Quantity<Quantity,Unit>:
      type: object
      properties:
        amount:
          type: number
        unit:
          $ref: '#/components/schemas/Unit'
    Unit:
      type: object
    Time:
      allOf:
      - $ref: '#/components/schemas/Quantity<Quantity,Unit>'
      - type: object
        properties:
          amount:
            type: number
  securitySchemes:
    basicAuth:
      type: http
      scheme: basic
```

**Comment by croudet** [comment]
Can you give a try to this branch https://github.com/croudet/micronaut-openapi/tree/Issue%23259

**Comment by bjonnhy** [comment]
Using your branch produced a much better result, thanks!

I still have some concerns around how the generation produces a lot of 'duplicates' when there are multiple subclasses. This could become somewhat 'noisy' in the specification.

Now produced result:
```yml
openapi: 3.0.1
info:
  title: api
  description: API
  version: "0.01"
servers:
- url: http://dev.lan
paths:
  /time:
    get:
      summary: Time of 10 seconds
      description: Time of 10 seconds
      operationId: getTime
      parameters: []
      responses:
        default:
          description: Time of 10 seconds
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/Time'
  /length:
    get:
      summary: Length of 10 Kilometer
      description: Length of 10 Kilometer
      operationId: getLength
      parameters: []
      responses:
        default:
          description: Length of 10 Kilometer
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/Length'
components:
  schemas:
    Quantity<Time,TimeUnit>:
      type: object
      properties:
        amount:
          type: number
        unit:
          $ref: '#/components/schemas/TimeUnit'
    TimeUnit:
      type: string
      enum:
      - Millisecond
      - Second
      - Minute
      - Hour
      - Day
      - Week
    Time:
      type: object
      allOf:
      - $ref: '#/components/schemas/Quantity<Time,TimeUnit>'
    Quantity<Length,LengthUnit>:
      type: object
      properties:
        amount:
          type: number
        unit:
          $ref: '#/components/schemas/LengthUnit'
    LengthUnit:
      type: string
      enum:
      - Millimeter
      - Centimeter
      - Decimeter
      - Meter
      - Kilometer
    Length:
      type: object
      allOf:
      - $ref: '#/components/schemas/Quantity<Length,LengthUnit>'
  securitySchemes:
    basicAuth:
      type: http
      scheme: basic
```
Less 'noisy' suggestion:
```yml
openapi: 3.0.1
info:
  title: api
  description: API
  version: "0.01"
servers:
- url: http://dev.lan
paths:
  /time:
    get:
      summary: Time of 10 seconds
      description: Time of 10 seconds
      operationId: getTime
      parameters: []
      responses:
        default:
          description: Time of 10 seconds
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/Time'
  /length:
    get:
      summary: Length of 10 Kilometer
      description: Length of 10 Kilometer
      operationId: getLength
      parameters: []
      responses:
        default:
          description: Length of 10 Kilometer
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/Length'
components:
  schemas:
    Quantity_Quantity.Unit_:
      type: object
      properties:
        amount:
          type: number
        unit:
          type: object
    Time:
      type: object
      properties:
        unit:
          $ref: '#/components/schemas/TimeUnit'
      allOf:
        - $ref: '#/components/schemas/Quantity_Quantity.Unit_'
    TimeUnit:
      type: string
      enum:
      - Millisecond
      - Second
      - Minute
      - Hour
      - Day
      - Week
    Length:
      type: object
      properties:
        unit:
          $ref: '#/components/schemas/LengthUnit'
      allOf:
        - $ref: '#/components/schemas/Quantity_Quantity.Unit_'
    LengthUnit:
      type: string
      enum:
      - Millimeter
      - Centimeter
      - Decimeter
      - Meter
      - Kilometer
  securitySchemes:
    basicAuth:
      type: http
      scheme: basic
```
Or dropping the `allOf` inheritance all together:
```yml
openapi: 3.0.1
info:
  title: api
  description: API
  version: "0.01"
servers:
- url: http://dev.lan
paths:
  /time:
    get:
      summary: Time of 10 seconds
      description: Time of 10 seconds
      operationId: getTime
      parameters: []
      responses:
        default:
          description: Time of 10 seconds
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/Time'
  /length:
    get:
      summary: Length of 10 Kilometer
      description: Length of 10 Kilometer
      operationId: getLength
      parameters: []
      responses:
        default:
          description: Length of 10 Kilometer
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/Length'
components:
  schemas:
    Time:
      type: object
      properties:
        amount:
          type: number
        unit:
          $ref: '#/components/schemas/TimeUnit'
    TimeUnit:
      type: string
      enum:
      - Millisecond
      - Second
      - Minute
      - Hour
      - Day
      - Week
    Length:
      type: object
      properties:
        amount:
          type: number
        unit:
          $ref: '#/components/schemas/LengthUnit'
    LengthUnit:
      type: string
      enum:
      - Millimeter
      - Centimeter
      - Decimeter
      - Meter
      - Kilometer
  securitySchemes:
    basicAuth:
      type: http
      scheme: basic
```

**Comment by alvarosanchez** [comment]
Fixed by #261 and #262
